File: backend/app/admin/views/email_campaigns.py
"""Email campaign admin view for SQLAdmin."""

from sqladmin import ModelView
from app.models.email import EmailCampaign


def _format_status(model, prop):
    """Format campaign status."""
    status_map = {
        "draft": "Draft",
        "sending": "Sending",
        "completed": "Completed",
        "failed": "Failed",
    }
    return status_map.get(model.status, model.status)


def _format_user(model, prop):
    """Format user relationship."""
    return str(model.user) if model.user else "-"


def _format_organization(model, prop):
    """Format organization relationship."""
    return str(model.organization) if model.organization else "Global"


def _format_template(model, prop):
    """Format template relationship."""
    return str(model.template) if model.template else "-"


def _format_domain(model, prop):
    """Format domain relationship."""
    return str(model.domain) if model.domain else "-"


def _format_stats(model, prop):
    """Format delivery statistics."""
    return f"{model.delivered_count}/{model.sent_count} delivered"


class EmailCampaignAdmin(ModelView, model=EmailCampaign):
    """Admin view for EmailCampaign model."""

    name = "Email Campaign"
    name_plural = "Email Campaigns"
    icon = "fa-solid fa-bullhorn"

    # Columns to display in list view
    column_list = [
        EmailCampaign.id,
        EmailCampaign.name,
        EmailCampaign.status,
        EmailCampaign.total_recipients,
        EmailCampaign.sent_count,
        EmailCampaign.delivered_count,
        EmailCampaign.opened_count,
        EmailCampaign.bounced_count,
        EmailCampaign.user,
        EmailCampaign.organization,
        EmailCampaign.created_at,
    ]

    # Columns to search
    column_searchable_list = [EmailCampaign.name, EmailCampaign.subject]

    # Default sort
    column_default_sort = [(EmailCampaign.created_at, True)]  # True = descending

    # Columns that are sortable
    column_sortable_list = [
        EmailCampaign.id,
        EmailCampaign.name,
        EmailCampaign.status,
        EmailCampaign.total_recipients,
        EmailCampaign.sent_count,
        EmailCampaign.delivered_count,
        EmailCampaign.opened_count,
        EmailCampaign.bounced_count,
        EmailCampaign.created_at,
    ]

    # Make certain fields read-only
    form_readonly_columns = [
        EmailCampaign.created_at,
        EmailCampaign.updated_at,
        EmailCampaign.started_at,
        EmailCampaign.completed_at,
        EmailCampaign.total_recipients,
        EmailCampaign.sent_count,
        EmailCampaign.delivered_count,
        EmailCampaign.bounced_count,
        EmailCampaign.opened_count,
        EmailCampaign.clicked_count,
        EmailCampaign.spam_count,
        EmailCampaign.failed_count,
    ]

    # Details view columns
    column_details_list = [
        EmailCampaign.id,
        EmailCampaign.name,
        EmailCampaign.subject,
        EmailCampaign.body,
        EmailCampaign.status,
        EmailCampaign.template,
        EmailCampaign.domain,
        EmailCampaign.from_email,
        EmailCampaign.from_name,
        EmailCampaign.total_recipients,
        EmailCampaign.sent_count,
        EmailCampaign.delivered_count,
        EmailCampaign.opened_count,
        EmailCampaign.clicked_count,
        EmailCampaign.bounced_count,
        EmailCampaign.spam_count,
        EmailCampaign.failed_count,
        EmailCampaign.search_result_ids,
        EmailCampaign.user,
        EmailCampaign.organization,
        EmailCampaign.created_at,
        EmailCampaign.updated_at,
        EmailCampaign.started_at,
        EmailCampaign.completed_at,
    ]

    # Column formatters for human-readable display
    column_formatters = {
        EmailCampaign.status: _format_status,
        EmailCampaign.user: _format_user,
        EmailCampaign.organization: _format_organization,
        EmailCampaign.template: _format_template,
        EmailCampaign.domain: _format_domain,
    }

    # Column labels for better readability
    column_labels = {
        EmailCampaign.name: "Name",
        EmailCampaign.subject: "Subject",
        EmailCampaign.body: "Body",
        EmailCampaign.status: "Status",
        EmailCampaign.template: "Template",
        EmailCampaign.domain: "Domain",
        EmailCampaign.from_email: "From Email",
        EmailCampaign.from_name: "From Name",
        EmailCampaign.total_recipients: "Recipients",
        EmailCampaign.sent_count: "Sent",
        EmailCampaign.delivered_count: "Delivered",
        EmailCampaign.opened_count: "Opened",
        EmailCampaign.clicked_count: "Clicked",
        EmailCampaign.bounced_count: "Bounced",
        EmailCampaign.spam_count: "Spam",
        EmailCampaign.failed_count: "Failed",
        EmailCampaign.search_result_ids: "Search Results",
        EmailCampaign.user: "Owner",
        EmailCampaign.organization: "Organization",
        EmailCampaign.started_at: "Started At",
        EmailCampaign.completed_at: "Completed At",
    }

    # Form columns - exclude relationships from form
    form_excluded_columns = [EmailCampaign.logs]

File: frontend/src/services/api/emailCampaigns.ts
/**
 * Email campaigns API client.
 */

import { apiClient } from '@/client';

export interface EmailCampaign {
  id: number;
  name: string;
  subject: string;
  status: 'draft' | 'sending' | 'completed' | 'failed';
  total_recipients: number;
  sent_count: number;
  delivered_count: number;
  opened_count: number;
  clicked_count: number;
  bounced_count: number;
  spam_count: number;
  failed_count: number;
  from_email?: string;
  from_name?: string;
  created_at: string;
  started_at?: string;
  completed_at?: string;
}

export interface EmailLog {
  id: number;
  campaign_id?: number;
  to_email: string;
  to_name?: string;
  subject: string;
  status: 'pending' | 'sent' | 'delivered' | 'bounced' | 'opened' | 'clicked' | 'spam' | 'failed';
  external_message_id?: string;
  error_message?: string;
  created_at: string;
  sent_at?: string;
  delivered_at?: string;
  opened_at?: string;
  clicked_at?: string;
  bounced_at?: string;
}

export interface CampaignStats {
  total: number;
  sent: number;
  delivered: number;
  opened: number;
  clicked: number;
  bounced: number;
  spam: number;
  failed: number;
  delivery_rate: number;
  open_rate: number;
  click_rate: number;
  bounce_rate: number;
}

export interface CampaignsListParams {
  limit?: number;
  offset?: number;
  status?: string;
}

export interface EmailReply {
  id: number;
  from_email: string;
  from_name: string | null;
  subject: string;
  body_text: string | null;
  campaign_id: number | null;
  is_processed: boolean;
  forwarded_to: string | null;
  received_at: string;
}

/**
 * List email campaigns.
 */
export async function listCampaigns(params: CampaignsListParams = {}): Promise<EmailCampaign[]> {
  const response = await apiClient.get<EmailCampaign[]>('/email/campaigns', { params });
  return response.data ?? [];
}

/**
 * Get a single campaign.
 */
export async function getCampaign(id: number): Promise<EmailCampaign> {
  const response = await apiClient.get<EmailCampaign>(`/email/campaigns/${id}`);
  return response.data;
}

/**
 * Get campaign statistics.
 */
export async function getCampaignStats(id: number): Promise<CampaignStats> {
  const response = await apiClient.get<CampaignStats>(`/email/campaigns/${id}/stats`);
  return response.data;
}

/**
 * Get campaign logs.
 */
export async function getCampaignLogs(
  id: number,
  params: { limit?: number; offset?: number; status?: string } = {}
): Promise<EmailLog[]> {
  const response = await apiClient.get<EmailLog[]>(`/email/campaigns/${id}/logs`, { params });
  return response.data ?? [];
}

/**
 * Get overall email statistics.
 */
export async function getEmailStats(): Promise<CampaignStats> {
  const response = await apiClient.get<CampaignStats>('/email/stats');
  return response.data;
}

/**
 * Get email replies.
 */
export async function getReplies(
  params: { limit?: number; offset?: number } = {}
): Promise<{ replies: EmailReply[]; total: number }> {
  const response = await apiClient.get<{ replies: EmailReply[]; total: number }>('/email/replies', { params });
  return response.data ?? { replies: [], total: 0 };
}

/**
 * Get a specific email reply.
 */
export async function getReply(id: number): Promise<EmailReply> {
  const response = await apiClient.get<EmailReply>(`/email/replies/${id}`);
  return response.data;
}
